crud/select/merger.lua

```diff
@@ -0,0 +1,235 @@
+local buffer = require('buffer')
+local msgpack = require('msgpack')
+local log = require('log')
+local ffi = require('ffi')
+local collations = require('crud.common.collations')
+local comparators = require('crud.select.comparators')
+
+local key_def_lib
+local merger_lib
+
+if pcall(require, 'tuple.merger') then
+    merger_lib = require('tuple.merger')
+    key_def_lib = require('tuple.keydef')
+elseif pcall(require, 'merger') then
+    log.info('Impossible to load "tuple-merger" module. Built-in "merger" is used')
+    merger_lib = require('merger')
+    key_def_lib = require('key_def')
+else
+    error('Seems your Tarantool version (' .. _TARANTOOL ..
+            ') does not support "tuple-merger" or "merger" modules')
+end
+
+local key_def_cache = {}
+setmetatable(key_def_cache, {__mode = 'k'})
+
+-- As "tuple.key_def" doesn't support collation_id
+-- we manually change it to collation
+local function normalize_parts(index_parts)
+    local result = {}
+
+    for _, part in ipairs(index_parts) do
+        if part.collation_id == nil then
+            table.insert(result, part)
+        else
+            local part_copy = table.copy(part)
+            part_copy.collation = collations.get(part)
+            part_copy.collation_id = nil
+            table.insert(result, part_copy)
+        end
+    end
+
+    return result
+end
+
+local function get_key_def(replicasets, space_name, field_names, index_name)
+    -- Get requested and primary index metainfo.
+    local conn = select(2, next(replicasets)).master.conn
+    local space = conn.space[space_name]
+    local space_format = space:format()
+    local index = space.index[index_name]
+    local index_by_field_names = {index, field_names}
+
+    if key_def_cache[index_by_field_names] ~= nil then
+        return key_def_cache[index_by_field_names]
+    end
```

Comment:
It looks a bit tricky. You need to calculate some hash (string). Otherwise this table will be recreated at each call - finally, it will work as without any cache.

Alternative - use `msgpack.encode(index_by_field_names)` as key.